In [1]:
import pandas as pd
import numpy as np
import re

# showing lots of rows and columns
pd.set_option('display.max_columns', 999)
pd.set_option('display.max_rows', 999)

# hiding annoying pink warning boxes
import warnings
warnings.filterwarnings("ignore")

In [2]:
hop_teaming_subset = pd.read_csv('data/Filtered_Hop_Team_Data.csv')

In [3]:
hop_teaming_subset.head()

Unnamed: 0    from_npi      to_npi  patient_count  transaction_count  \
0           0  1003819046  1023055126             11                 11   
1           1  1750384780  1023055126             48                 63   
2           2  1750384780  1023055126             48                 63   
3           3  1750384780  1396882205             24                 58   
4           4  1750384780  1396882205             24                 58   

   average_day_wait    Phys_NPI  Phys_Office Phys_Last_Name Phys_First_Name  \
0            13.182  1003819046          NaN       NYLANDER         BARBARA   
1            36.921  1750384780          NaN       PERRIGIN           JULIE   
2            36.921  1750384780          NaN       PERRIGIN           JULIE   
3            32.759  1750384780          NaN       PERRIGIN           JULIE   
4            32.759  1750384780          NaN       PERRIGIN           JULIE   

                Phys_Class   Phys_Spec    Hosp_NPI  \
0  Obstetrics & Gynecology  Gynecology  1023055126   
1          Family Medicine         NaN  1023055126   
2          Family Medicine         NaN  1023055126   
3          Family Medicine         NaN  1396882205   
4          Family Medicine         NaN  1396882205   

                                 Hospital                   Hosp_Class  \
0  HCA HEALTH SERVICES OF TENNESSEE, INC.  General Acute Care Hospital   
1  HCA HEALTH SERVICES OF TENNESSEE, INC.  General Acute Care Hospital   
2  HCA HEALTH SERVICES OF TENNESSEE, INC.  General Acute Care Hospital   
3    VANDERBILT UNIVERSITY MEDICAL CENTER  General Acute Care Hospital   
4    VANDERBILT UNIVERSITY MEDICAL CENTER  General Acute Care Hospital   

  Hosp_Spec  
0       NaN  
1       NaN  
2       NaN  
3       NaN  
4       NaN

In [4]:
hop_teaming_subset.shape

(35731, 16)

### Additional Data Cleaning Steps

In [5]:
# drop duplicate index
hop_teaming_subset.drop('Unnamed: 0', axis = 1, inplace = True)
hop_teaming_subset.head()

from_npi      to_npi  patient_count  transaction_count  average_day_wait  \
0  1003819046  1023055126             11                 11            13.182   
1  1750384780  1023055126             48                 63            36.921   
2  1750384780  1023055126             48                 63            36.921   
3  1750384780  1396882205             24                 58            32.759   
4  1750384780  1396882205             24                 58            32.759   

     Phys_NPI  Phys_Office Phys_Last_Name Phys_First_Name  \
0  1003819046          NaN       NYLANDER         BARBARA   
1  1750384780          NaN       PERRIGIN           JULIE   
2  1750384780          NaN       PERRIGIN           JULIE   
3  1750384780          NaN       PERRIGIN           JULIE   
4  1750384780          NaN       PERRIGIN           JULIE   

                Phys_Class   Phys_Spec    Hosp_NPI  \
0  Obstetrics & Gynecology  Gynecology  1023055126   
1          Family Medicine         NaN  1023055126   
2          Family Medicine         NaN  1023055126   
3          Family Medicine         NaN  1396882205   
4          Family Medicine         NaN  1396882205   

                                 Hospital                   Hosp_Class  \
0  HCA HEALTH SERVICES OF TENNESSEE, INC.  General Acute Care Hospital   
1  HCA HEALTH SERVICES OF TENNESSEE, INC.  General Acute Care Hospital   
2  HCA HEALTH SERVICES OF TENNESSEE, INC.  General Acute Care Hospital   
3    VANDERBILT UNIVERSITY MEDICAL CENTER  General Acute Care Hospital   
4    VANDERBILT UNIVERSITY MEDICAL CENTER  General Acute Care Hospital   

  Hosp_Spec  
0       NaN  
1       NaN  
2       NaN  
3       NaN  
4       NaN

In [6]:
# make all column names lowercase
hop_teaming_subset.columns = map(str.lower, hop_teaming_subset.columns)
hop_teaming_subset.head()

from_npi      to_npi  patient_count  transaction_count  average_day_wait  \
0  1003819046  1023055126             11                 11            13.182   
1  1750384780  1023055126             48                 63            36.921   
2  1750384780  1023055126             48                 63            36.921   
3  1750384780  1396882205             24                 58            32.759   
4  1750384780  1396882205             24                 58            32.759   

     phys_npi  phys_office phys_last_name phys_first_name  \
0  1003819046          NaN       NYLANDER         BARBARA   
1  1750384780          NaN       PERRIGIN           JULIE   
2  1750384780          NaN       PERRIGIN           JULIE   
3  1750384780          NaN       PERRIGIN           JULIE   
4  1750384780          NaN       PERRIGIN           JULIE   

                phys_class   phys_spec    hosp_npi  \
0  Obstetrics & Gynecology  Gynecology  1023055126   
1          Family Medicine         NaN  1023055126   
2          Family Medicine         NaN  1023055126   
3          Family Medicine         NaN  1396882205   
4          Family Medicine         NaN  1396882205   

                                 hospital                   hosp_class  \
0  HCA HEALTH SERVICES OF TENNESSEE, INC.  General Acute Care Hospital   
1  HCA HEALTH SERVICES OF TENNESSEE, INC.  General Acute Care Hospital   
2  HCA HEALTH SERVICES OF TENNESSEE, INC.  General Acute Care Hospital   
3    VANDERBILT UNIVERSITY MEDICAL CENTER  General Acute Care Hospital   
4    VANDERBILT UNIVERSITY MEDICAL CENTER  General Acute Care Hospital   

  hosp_spec  
0       NaN  
1       NaN  
2       NaN  
3       NaN  
4       NaN

In [7]:
# quickly checking column dtypes for additional changes that may need to be made
hop_teaming_subset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 35731 entries, 0 to 35730
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   from_npi           35731 non-null  int64  
 1   to_npi             35731 non-null  int64  
 2   patient_count      35731 non-null  int64  
 3   transaction_count  35731 non-null  int64  
 4   average_day_wait   35731 non-null  float64
 5   phys_npi           35731 non-null  int64  
 6   phys_office        0 non-null      float64
 7   phys_last_name     35725 non-null  object 
 8   phys_first_name    35731 non-null  object 
 9   phys_class         35731 non-null  object 
 10  phys_spec          18064 non-null  object 
 11  hosp_npi           35731 non-null  int64  
 12  hospital           35731 non-null  object 
 13  hosp_class         35731 non-null  object 
 14  hosp_spec          1388 non-null   object 
dtypes: float64(2), int64(6), object(7)
memory usage: 4.1+ MB


**Note:** We have a lot of missing values for phy_office (all values are missing), phys_spec, and hosp_spec...what might have happened here?

### Data Exploration  
  
#### Q1a. Build a profile of providers referring patients to the major hospitals in Nashville.  

In [8]:
hop_teaming_subset

from_npi      to_npi  patient_count  transaction_count  \
0      1003819046  1023055126             11                 11   
1      1750384780  1023055126             48                 63   
2      1750384780  1023055126             48                 63   
3      1750384780  1396882205             24                 58   
4      1750384780  1396882205             24                 58   
...           ...         ...            ...                ...   
35726  1558887091  1164590386             17                 17   
35727  1760908362  1447571658             13                 15   
35728  1306362637  1780778969             21                 21   
35729  1063923407  1396882205             13                 13   
35730  1265943328  1861479545             12                 14   

       average_day_wait    phys_npi  phys_office phys_last_name  \
0                13.182  1003819046          NaN       NYLANDER   
1                36.921  1750384780          NaN       PERRIGIN   
2                36.921  1750384780          NaN       PERRIGIN   
3                32.759  1750384780          NaN       PERRIGIN   
4                32.759  1750384780          NaN       PERRIGIN   
...                 ...         ...          ...            ...   
35726             7.294  1558887091          NaN         DUNCAN   
35727            12.200  1760908362          NaN          BLATT   
35728             1.000  1306362637          NaN        WILGING   
35729             7.231  1063923407          NaN        PICKENS   
35730             9.000  1265943328          NaN           BASS   

      phys_first_name                               phys_class   phys_spec  \
0             BARBARA                  Obstetrics & Gynecology  Gynecology   
1               JULIE                          Family Medicine         NaN   
2               JULIE                          Family Medicine         NaN   
3               JULIE                          Family Medicine         NaN   
4               JULIE                          Family Medicine         NaN   
...               ...                                      ...         ...   
35726         JEFFREY  Nurse Anesthetist, Certified Registered         NaN   
35727       KATHERINE                       Physical Therapist         NaN   
35728            ERIN  Nurse Anesthetist, Certified Registered         NaN   
35729         REBECCA                       Nurse Practitioner  Acute Care   
35730        JONATHAN                       Physical Therapist         NaN   

         hosp_npi                                hospital  \
0      1023055126  HCA HEALTH SERVICES OF TENNESSEE, INC.   
1      1023055126  HCA HEALTH SERVICES OF TENNESSEE, INC.   
2      1023055126  HCA HEALTH SERVICES OF TENNESSEE, INC.   
3      1396882205    VANDERBILT UNIVERSITY MEDICAL CENTER   
4      1396882205    VANDERBILT UNIVERSITY MEDICAL CENTER   
...           ...                                     ...   
35726  1164590386        SAINT THOMAS RUTHERFORD HOSPITAL   
35727  1447571658      SUMNER REGIONAL MEDICAL CENTER LLC   
35728  1780778969              SAINT THOMAS WEST HOSPITAL   
35729  1396882205    VANDERBILT UNIVERSITY MEDICAL CENTER   
35730  1861479545                 MAURY REGIONAL HOSPITAL   

                        hosp_class hosp_spec  
0      General Acute Care Hospital       NaN  
1      General Acute Care Hospital       NaN  
2      General Acute Care Hospital       NaN  
3      General Acute Care Hospital       NaN  
4      General Acute Care Hospital       NaN  
...                            ...       ...  
35726  General Acute Care Hospital       NaN  
35727  General Acute Care Hospital       NaN  
35728  General Acute Care Hospital       NaN  
35729  General Acute Care Hospital       NaN  
35730  General Acute Care Hospital       NaN  

[35731 rows x 15 columns]

In [9]:
# check out which facilities/hospitals are represented in the dataset (in alphabetical order)
hop_teaming_subset['hospital'].sort_values().unique()

array(['CENTRAL TENNESSEE HOSPITAL CORPORATION',
       'HCA HEALTH SERVICES OF TENNESSEE, INC.',
       'HENDERSONVILLE HOSPITAL CORPORATION',
       'HTI MEMORIAL HOSPITAL CORPORATION',
       'MACON COUNTY GENERAL HOSPITAL, INC.', 'MAURY REGIONAL HOSPITAL',
       'NASHVILLE GENERAL HOSPITAL', 'NORTHCREST MEDICAL CENTER',
       'RIVER PARK HOSPITAL LLC', 'RIVERVIEW MEDICAL CENTER LLC',
       'SAINT THOMAS DEKALB HOSPITAL, LLC',
       'SAINT THOMAS RIVER PARK HOSPITAL, LLC',
       'SAINT THOMAS RUTHERFORD HOSPITAL',
       'SAINT THOMAS STONES RIVER HOSPITAL, LLC',
       'SAINT THOMAS WEST HOSPITAL', 'SETON CORPORATION',
       'SHELBYVILLE HOSPITAL COMPANY LLC',
       'SMITH COUNTY MEMORIAL HOSPITAL LLC',
       'SUMNER REGIONAL HEALTH SYSTEMS, INC',
       'SUMNER REGIONAL MEDICAL CENTER LLC',
       'TROUSDALE MEDICAL CENTER LLC', 'TRUSTPOINT HOSPITAL, LLC',
       'VANDERBILT UNIVERSITY MEDICAL CENTER',
       'WILLIAMSON COUNTY HOSPITAL DISTRICT'], dtype=object)

#### Notes:  
1. These results not only represent specific facilities, but also networks of facilities.  
For example, HCA Health Services of Tennessee, Inc. is comprised of at least 10 separate facilties in Nashville (based on the information available at https://hcahealthcare.com/locations/?state=TN):   
*The Children's Hospital at TriStar Centennial  
*Ashland City Medical Center  
*Centennial Medical Center  
*Centennial Parthenon Pavilion  
*Hendersonville Medical Center  
*Horizon Medical Center  
*Skyline Medical Center  
*Southern Hills Medical Center  
*StoneCrest Medical Center  
*Summit Medical Center  
  
I'm not yet sure if retaining the zipcodes/Address_Postal_Codes column in our SQL query could help us distinguish between different locations--or if having facilities broken out in this way matters for our analyses. Either way, we need to take this reality into account as we report on our findings.  
  
2. It may be better to treat "Seton Corporation" as "Saint Thomas West Hospital" since both of their NPIs are associated with the same address...(based on the info available at https://npino.com/hospital/1992861314-seton-corporation/)   
  
3. Similarly, it may be better to treat "Central Tennessee Hospital Corporation" as "HCA Health Services of Tennessee, Inc." for similar reasons. "Central Tennessee Hospital Corporation" is apparently this facility's legal business name, but it does business as "Horizon Medical Center", which is within "HCA Health Services of Tennessee, Inc." (https://npidb.org/organizations/hospitals/general-acute-care-hospital_282n00000x/1598702557.aspx and https://www.bloomberg.com/profile/company/0123652D:US)

In [10]:
# count number of rows/to-from relationships associated with each facility (in alphabetical order same as above)
hop_teaming_subset['hospital'].value_counts().sort_index(ascending = True)

CENTRAL TENNESSEE HOSPITAL CORPORATION      838
HCA HEALTH SERVICES OF TENNESSEE, INC.     7488
HENDERSONVILLE HOSPITAL CORPORATION        1274
HTI MEMORIAL HOSPITAL CORPORATION          1743
MACON COUNTY GENERAL HOSPITAL, INC.         364
MAURY REGIONAL HOSPITAL                    1832
NASHVILLE GENERAL HOSPITAL                  283
NORTHCREST MEDICAL CENTER                   759
RIVER PARK HOSPITAL LLC                       2
RIVERVIEW MEDICAL CENTER LLC                299
SAINT THOMAS DEKALB HOSPITAL, LLC           129
SAINT THOMAS RIVER PARK HOSPITAL, LLC       558
SAINT THOMAS RUTHERFORD HOSPITAL           1666
SAINT THOMAS STONES RIVER HOSPITAL, LLC     117
SAINT THOMAS WEST HOSPITAL                 5963
SETON CORPORATION                             1
SHELBYVILLE HOSPITAL COMPANY LLC            378
SMITH COUNTY MEMORIAL HOSPITAL LLC            1
SUMNER REGIONAL HEALTH SYSTEMS, INC          24
SUMNER REGIONAL MEDICAL CENTER LLC         1112
TROUSDALE MEDICAL CENTER LLC            

In [11]:
# count referral locations (from highest to lowest)
hop_teaming_subset['hospital'].value_counts()

VANDERBILT UNIVERSITY MEDICAL CENTER       8629
HCA HEALTH SERVICES OF TENNESSEE, INC.     7488
SAINT THOMAS WEST HOSPITAL                 5963
WILLIAMSON COUNTY HOSPITAL DISTRICT        2117
MAURY REGIONAL HOSPITAL                    1832
HTI MEMORIAL HOSPITAL CORPORATION          1743
SAINT THOMAS RUTHERFORD HOSPITAL           1666
HENDERSONVILLE HOSPITAL CORPORATION        1274
SUMNER REGIONAL MEDICAL CENTER LLC         1112
CENTRAL TENNESSEE HOSPITAL CORPORATION      838
NORTHCREST MEDICAL CENTER                   759
SAINT THOMAS RIVER PARK HOSPITAL, LLC       558
SHELBYVILLE HOSPITAL COMPANY LLC            378
MACON COUNTY GENERAL HOSPITAL, INC.         364
RIVERVIEW MEDICAL CENTER LLC                299
NASHVILLE GENERAL HOSPITAL                  283
TROUSDALE MEDICAL CENTER LLC                141
SAINT THOMAS DEKALB HOSPITAL, LLC           129
SAINT THOMAS STONES RIVER HOSPITAL, LLC     117
SUMNER REGIONAL HEALTH SYSTEMS, INC          24
TRUSTPOINT HOSPITAL, LLC                

In [12]:
# what percentage of referrals do these facilities receive overall from all providers in our subset?
hop_teaming_subset['hospital'].value_counts(normalize = True) # normalize = True yields value counts as percentages

VANDERBILT UNIVERSITY MEDICAL CENTER       0.241499
HCA HEALTH SERVICES OF TENNESSEE, INC.     0.209566
SAINT THOMAS WEST HOSPITAL                 0.166886
WILLIAMSON COUNTY HOSPITAL DISTRICT        0.059248
MAURY REGIONAL HOSPITAL                    0.051272
HTI MEMORIAL HOSPITAL CORPORATION          0.048781
SAINT THOMAS RUTHERFORD HOSPITAL           0.046626
HENDERSONVILLE HOSPITAL CORPORATION        0.035655
SUMNER REGIONAL MEDICAL CENTER LLC         0.031121
CENTRAL TENNESSEE HOSPITAL CORPORATION     0.023453
NORTHCREST MEDICAL CENTER                  0.021242
SAINT THOMAS RIVER PARK HOSPITAL, LLC      0.015617
SHELBYVILLE HOSPITAL COMPANY LLC           0.010579
MACON COUNTY GENERAL HOSPITAL, INC.        0.010187
RIVERVIEW MEDICAL CENTER LLC               0.008368
NASHVILLE GENERAL HOSPITAL                 0.007920
TROUSDALE MEDICAL CENTER LLC               0.003946
SAINT THOMAS DEKALB HOSPITAL, LLC          0.003610
SAINT THOMAS STONES RIVER HOSPITAL, LLC    0.003274
SUMNER REGIO

In [13]:
# How many unique referring providers are in our subset
hop_teaming_subset['from_npi'].nunique()

6594

It looks like there are only 6594 unique values for from_npi...is this right? Is there a way to make sure we didn't lose/accidentally drop some providers when we subset our data using sqlite?

In [14]:
# using groupby to see proportions of referrals to different facilities by provider
hop_teaming_subset.groupby('from_npi')['hospital'].value_counts(normalize = True)

from_npi    hospital                              
1003013160  HCA HEALTH SERVICES OF TENNESSEE, INC.    0.333333
            SAINT THOMAS WEST HOSPITAL                0.333333
            VANDERBILT UNIVERSITY MEDICAL CENTER      0.333333
1003013947  HCA HEALTH SERVICES OF TENNESSEE, INC.    0.666667
            VANDERBILT UNIVERSITY MEDICAL CENTER      0.333333
                                                        ...   
1992985949  SAINT THOMAS WEST HOSPITAL                0.333333
            VANDERBILT UNIVERSITY MEDICAL CENTER      0.166667
            WILLIAMSON COUNTY HOSPITAL DISTRICT       0.166667
1992987085  VANDERBILT UNIVERSITY MEDICAL CENTER      1.000000
1992998439  VANDERBILT UNIVERSITY MEDICAL CENTER      1.000000
Name: hospital, Length: 23175, dtype: float64

How do I get this above output into a dataframe that can be more fully explored? If I could do this, I think it gives us at least a good start on our way to answering this first question about a profile of providers referring patients to hospitals in Nashville.

In [18]:
# create a groupby object
provider_profile = hop_teaming_subset.groupby(['from_npi', 'hospital'])
provider_profile

In [21]:
provider_profile.size()

from_npi    hospital                              
1003013160  HCA HEALTH SERVICES OF TENNESSEE, INC.    1
            SAINT THOMAS WEST HOSPITAL                1
            VANDERBILT UNIVERSITY MEDICAL CENTER      1
1003013947  HCA HEALTH SERVICES OF TENNESSEE, INC.    2
            VANDERBILT UNIVERSITY MEDICAL CENTER      1
                                                     ..
1992985949  SAINT THOMAS WEST HOSPITAL                2
            VANDERBILT UNIVERSITY MEDICAL CENTER      1
            WILLIAMSON COUNTY HOSPITAL DISTRICT       1
1992987085  VANDERBILT UNIVERSITY MEDICAL CENTER      2
1992998439  VANDERBILT UNIVERSITY MEDICAL CENTER      1
Length: 23175, dtype: int64

This isn't an improvement on the output in cell 14...

#### Q1b. Are certain specialties more likely to refer to a particular hospital over the others?  

In [22]:
# using groupby to see whether certain specialties refer patients to some facilities more than others
# note: this is where having so much missing info for specialities could bite us in the ass...
hop_teaming_subset.groupby('phys_spec')['hospital'].value_counts(normalize = True)

phys_spec       hospital                              
Acute Care      VANDERBILT UNIVERSITY MEDICAL CENTER      0.507666
                SAINT THOMAS WEST HOSPITAL                0.139693
                HCA HEALTH SERVICES OF TENNESSEE, INC.    0.126065
                MAURY REGIONAL HOSPITAL                   0.035775
                SAINT THOMAS RUTHERFORD HOSPITAL          0.035775
                                                            ...   
Women's Health  NORTHCREST MEDICAL CENTER                 0.018519
                RIVERVIEW MEDICAL CENTER LLC              0.018519
                SUMNER REGIONAL MEDICAL CENTER LLC        0.018519
                CENTRAL TENNESSEE HOSPITAL CORPORATION    0.009259
                SAINT THOMAS RUTHERFORD HOSPITAL          0.009259
Name: hospital, Length: 1025, dtype: float64

In [23]:
# what are the provider specialities provided in this subset?
hop_teaming_subset['phys_spec'].unique()

array(['Gynecology', nan, 'Cytopathology', 'Cardiovascular Disease',
       'Advanced Heart Failure and Transplant Cardiology',
       'Nuclear Radiology', 'Diagnostic Radiology',
       'Endocrinology, Diabetes & Metabolism',
       'Orthopaedic Surgery of the Spine', 'Neurology',
       'Clinical Cardiac Electrophysiology',
       'Interventional Pain Medicine', 'Pain Medicine', "Women's Health",
       'MOHS-Micrographic Surgery',
       'Ophthalmic Plastic and Reconstructive Surgery',
       'Pulmonary Disease', 'Family', 'Sleep Medicine', 'Acute Care',
       'Primary Podiatric Medicine', 'Nephrology', 'Geriatric Medicine',
       'Psychiatry', 'Sports Medicine', 'Vascular Surgery',
       'Gastroenterology', 'Vascular & Interventional Radiology',
       'Occupational Medicine', 'Pediatric Emergency Medicine',
       'Hand Surgery', 'Trauma Surgery', 'Medical',
       'Clinical Genetics (M.D.)', 'Gynecologic Oncology',
       'Anatomic Pathology & Clinical Pathology', 'Surgery of 

In [24]:
# how many distinct specialties are present in this subset?
hop_teaming_subset['phys_spec'].nunique()

118

In [26]:
# trying to get a peek at this info
hop_teaming_subset.groupby('phys_spec')['hospital'].value_counts(normalize = True).head(50)

phys_spec                       hospital                               
Acute Care                      VANDERBILT UNIVERSITY MEDICAL CENTER       0.507666
                                SAINT THOMAS WEST HOSPITAL                 0.139693
                                HCA HEALTH SERVICES OF TENNESSEE, INC.     0.126065
                                MAURY REGIONAL HOSPITAL                    0.035775
                                SAINT THOMAS RUTHERFORD HOSPITAL           0.035775
                                HTI MEMORIAL HOSPITAL CORPORATION          0.034072
                                WILLIAMSON COUNTY HOSPITAL DISTRICT        0.032368
                                CENTRAL TENNESSEE HOSPITAL CORPORATION     0.025554
                                HENDERSONVILLE HOSPITAL CORPORATION        0.015332
                                SAINT THOMAS RIVER PARK HOSPITAL, LLC      0.010221
                                SUMNER REGIONAL MEDICAL CENTER LLC         0.010221
    

Although I can get a glimpse at what's going on here, I still run into the same issue as above--I don't know how to convert this to a dataframe so that it's easier to explore these results.